### PR TITLE
Fix an issue where we are not injecting more HTTP2 connections when we should

### LINKED ISSUE #55980: [Regression] Multiple HTTP/2 streams created concurrently not all reaching the server
A gRPC test that works in .NET 5 now fails in .NET 6.

The test starts 201 gRPC calls over HTTP/2 concurrently, waits for them all to reach the server, and then completes. `SocketsHttpHandler.EnableMultipleHttp2Connections` is true so the client should be creating 3 HTTP/2 connections (server max streams is the default of 100) so all calls reach the server.

Logs seem to indicate that two connections are made and 200 calls reach the server. The 201st never makes it.

Attached are complete logs (LONG): [MaxConcurrentStreams_StartConcurrently.txt](https://github.com/dotnet/runtime/files/6845229/MaxConcurrentStreams_StartConcurrently.txt)

Repo: https://github.com/JamesNK/grpc-dotnet/tree/jamesnk/net6-concurrentstreams
Test: StreamingTests.MaxConcurrentStreams_StartConcurrently_AdditionalConnectionsCreated

https://github.com/JamesNK/grpc-dotnet/blob/5f9fe9a0442d99e3bce30be8b2ccf5ddbebd5beb/test/FunctionalTests/Client/StreamingTests.cs#L894-L992

```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-rc.1.21368.1
 Commit:    a5889518cc
```

### PR DESCRIPTION
**Pull request by geoffkizer**
Fixes #55980 

This is a regression caused by my connection pooling changes.

For HTTP2, we constrain our connection injection logic to never inject more than one connection at a time. On successful connection establishment, we should check if we need to create yet another connection to handle the queued requests. Unfortunately, we aren't doing this today.

The specific problem is a scenario like this:

(1) User submits one request. We enqueue it and kick off an HTTP2 connection attempt.
(2) User submits 100 more requests. There are now 101 requests in the queue.
(3) Connection attempt succeeds. We dequeue 100 requests and give them to the new connection to be processed. 
(4) There's still one request in the queue, so we should be kicking off another connection attempt -- but currently we don't.

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs [hunk 1]
```diff
@@ -2118,6 +2118,63 @@ public async Task Http2_MultipleConnectionsEnabled_ConnectionLimitNotReached_Con
...
+            using Http2LoopbackConnection c3 = await server.EstablishConnectionAsync(new SettingsEntry { SettingId = SettingId.MaxConcurrentStreams, Value = 100 });
+            (int finalStreamId, _) = await c3.ReadAndParseRequestHeaderAsync();
+            acceptedRequests.Add((c3, finalStreamId));
+
+            foreach (var request in acceptedRequests)
+            {
+                await request.connection.SendDefaultResponseAsync(request.streamId);
+            }
+
+            foreach (var t in sendTasks)
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:2171]: Nit:
```suggestion
            foreach (Task<HttpResponseMessage> t in sendTasks)
```

File: src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs [hunk 2]
```diff
@@ -2118,6 +2118,63 @@ public async Task Http2_MultipleConnectionsEnabled_ConnectionLimitNotReached_Con
...
+            {
+                (int streamId, _) = await c2.ReadAndParseRequestHeaderAsync();
+                acceptedRequests.Add((c2, streamId));
+            }
+
+            using Http2LoopbackConnection c3 = await server.EstablishConnectionAsync(new SettingsEntry { SettingId = SettingId.MaxConcurrentStreams, Value = 100 });
+            (int finalStreamId, _) = await c3.ReadAndParseRequestHeaderAsync();
+            acceptedRequests.Add((c3, finalStreamId));
+
+            foreach (var request in acceptedRequests)
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:2166]: Nit:
```suggestion
            foreach ((Http2LoopbackConnection connection, int streamId) request in acceptedRequests)
```

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
The outerloop System.Buffers failures will be fixed by https://github.com/dotnet/runtime/pull/56068.  Sorry about that.

**Comment by geoffkizer** [comment]
No worries. I doubt it will be the only outerloop failure :)

**Review by wfurt — APPROVED** [review]
LGTM

**Pull request merged**